### PR TITLE
forbid resource creation when APIBinding is deleting

### PR DESCRIPTION
**Pull request by qiujian16**
## Summary

follow up PR of https://github.com/kcp-dev/kcp/pull/1040
## Related issue(s)

Fixes https://github.com/kcp-dev/kcp/issues/846

### CODE REVIEW

File: pkg/server/apiextensions.go [hunk 1]
```diff
@@ -382,6 +382,31 @@ func shallowCopyCRD(in *apiextensionsv1.CustomResourceDefinition) *apiextensions
...
+func decorateCRDWithBinding(in *apiextensionsv1.CustomResourceDefinition, identity string, deleteTime *metav1.Time) *apiextensionsv1.CustomResourceDefinition {
+	out := shallowCopyCRD(in)
+
+	out.Annotations[apisv1alpha1.AnnotationAPIIdentityKey] = identity
+
+	if deleteTime.IsZero() {
+		return out
+	}
+
+	out.Status.Conditions = make([]apiextensionsv1.CustomResourceDefinitionCondition, len(in.Status.Conditions), len(in.Status.Conditions)+1)
```
Comment (stevekuznetsov) [inline pkg/server/apiextensions.go:397]: The slice does not need to grow if there's already a terminating condition, right?

File: pkg/server/apiextensions.go [hunk 2]
```diff
@@ -382,6 +382,31 @@ func shallowCopyCRD(in *apiextensionsv1.CustomResourceDefinition) *apiextensions
...
+
+	out.Status.Conditions = append(out.Status.Conditions, in.Status.Conditions...)
+
+	out.DeletionTimestamp = deleteTime.DeepCopy()
+	apiextensionshelpers.SetCRDCondition(out, apiextensionsv1.CustomResourceDefinitionCondition{
```
Comment (stevekuznetsov) [inline pkg/server/apiextensions.go:402]: This will take care of appending a new member to the slice if needed, right?

**Review by stevekuznetsov — COMMENTED** [review]
Are we happy with the racy implication of requiring APIBinding deletion to propagate to another object in order to stop creation?

File: pkg/server/apiextensions.go (hunk 1 above)
Comment (sttts) [inline pkg/server/apiextensions.go:397]: these conditions here are not visible to anybody, are they? They are only faked while feeding the apiextensions-apiserver CR registry.

Comment (sttts) [reply pkg/server/apiextensions.go:397]: nvmd. The Terminating condition is used by apiextensions. Odd. Maybe add a comment to clarify that. Using a condition always feel odd to me for any kind of critical logic.

File: pkg/server/apiextensions.go (hunk 2 above)
Comment (qiujian16) [reply pkg/server/apiextensions.go:404]: removed the setting of capacity.

**Comment by stevekuznetsov** [comment]
@sttts for CRDs? If it is, seems reasonable to me :)

**Comment by sttts** [comment]
In the sense that it is informer based in kube. So it can race.

File: pkg/reconciler/tenancy/clusterworkspacedeletion/clusterworkspace_deletion_controller.go [hunk 3]
```diff
@@ -242,6 +229,28 @@ func (c *Controller) patchCondition(ctx context.Context, old, new *tenancyv1alph
...
+	}
+
+	patch, err := json.Marshal(finalizerData)
+	if err != nil {
+		return err
+	}
+
+	// remove finalizers field if there is no remaining finalizers,
+	if len(finalizers) == 0 {
+		patch = []byte("{\"metadata\": {\"finalizers\": []}}")
```
Comment (sttts) [inline pkg/reconciler/tenancy/clusterworkspacedeletion/clusterworkspace_deletion_controller.go:246]: not sure this is safe. Removing the field is not really necessary. Just merge-patch your finalizer out. This patch here has the potential to override other finalizers just added. 

same for the same func above.

Comment (qiujian16) [reply pkg/reconciler/tenancy/clusterworkspacedeletion/clusterworkspace_deletion_controller.go]: added uid/rv as precond in this case. The finalizer is opmitempty, so I think we have to build this json string explicitly when finalizer is empty.

File: pkg/reconciler/tenancy/clusterworkspacedeletion/clusterworkspace_deletion_controller.go [hunk 4]
```diff
@@ -242,6 +229,31 @@ func (c *Controller) patchCondition(ctx context.Context, old, new *tenancyv1alph
...
+
+	patch, err := json.Marshal(finalizerData)
+	if err != nil {
+		return err
+	}
+
+	// remove finalizers field if there is no remaining finalizers,
+	if len(finalizers) == 0 {
+		patch = []byte(
+			fmt.Sprintf("{\"metadata\": {\"finalizers\": [], \"uid\": %q, \"resourceVersion\": %q}}", ws.UID, ws.ResourceVersion))
```
Comment (sttts) [inline pkg/reconciler/tenancy/clusterworkspacedeletion/clusterworkspace_deletion_controller.go:249]: At this point it's much easier to reason about a simple Update call with `ws` from above with the changes directly in the object.

Comment (qiujian16) [reply pkg/reconciler/tenancy/clusterworkspacedeletion/clusterworkspace_deletion_controller.go]: updated

**Review by sttts — APPROVED** [review]

**Pull request merged**